docs/source/deprecation.md

```diff
@@ -1,14 +1,164 @@
+has been merged into the upstream, but has been removed before the release
+or was marked as experimental all the time, is going to be removed without
+the deprecation state). The time period during which the deprecated entity
+won't be removed is at least 6 months. That doesn't mean we will remove
+everything deprecated immadiately after the 6 months, but it's to be expected
+that it will be dropped anytime between 6 and 12 months since the deprecation.
+
+In case of issues, deprecated functionality is not going to be fixed since
+it is deprecated (unless it is fixed e.g. as a side-effect of another
+problem fix).
+
+## How does leapp informs about deprecated functionality?
+
+Mainly via release notes and changelogs. In the official leapp related projects
+(especially leapp and leapp-repository) the OAMG team take care about release
+notes to ensure they contain all data about dropped and deprecated
+functionality.
+
+Additionally, when using leapp or snactor, user is notified via messages about
+deprecated entities in **limited cases** (see below). In case of the leapp
+utility, such messages are presented inside the generated reports. In case
+of the snactor utility, the information message is printed in the console
+output in the end of the snactor execution. See examples in this page for
+detail.
+
+But be aware of limitations of our detection in leapp and snactor utilities.
+Our detection mechanisms are not bulletproof. Here is list of some examples,
+when our detection does not work well:
+
+- when an actor a message of a deprecated model, the usage of the deprecated
+  model is detected only if the message has been produced before; IOW, in
+  case the message X is produced only on machines with a specific package,
+  the deprecated usage is not detected on machines without the packagw
+- when a message of a deprecated model is consumed, the message points
+  internals of the leapp framework, instead of place where the message is
+  consumed, e.g.
+  ```
+  Usage of deprecated Model "Foo" @ /usr/lib/python2.7/site-packages/leapp/models/__init__.py:113
+  Near:         return cls(init_method='to_model', **data)
+  ```
+- a deprecated class (including model) is reported only when an instance of the
+  class (object) is created during the runtime
+- use of a deprecated function/method is reported only when the function is
+  called
+
+There are probably other cases, but these are the most common, that could
+confuse people.
+
+
+## How to deprecate entities in leapp
+
+When you want to deprecate an entity in leapp projects, use the `deprecated`
+decorator from `leapp.utils.deprecation` above the definition of the entity.
+The decorator has two input parameters to be able to specify the start
+of the deprecation protection period and a message explaining that
+particular deprecation. E.g. in case the deprecated functionality has
+a replacement, it is expected it will be mentioned in the msg.
```

Comment:
code is better than a thousand words:) and link to the code is even better, so we have a single source of the truth. Let;s point to https://github.com/oamg/leapp/tree/master/tests/data/deprecation-tests

Reply:
Code is below. And our test cases are really not best guidance always ;-)